crypto/ocsp/ocsp_test.cc

```diff
@@ -512,114 +845,242 @@ static bssl::UniquePtr<X509> LoadX509fromPEM(const char *pem) {
-  ASSERT_TRUE(cert_id);
+  // Check valid connection timestamp right after OCSP response was validated.
+  connection_time = valid_after_ocsp_update_time;
+  ASSERT_EQ(-1, X509_cmp_time(thisupd, &connection_time));
+  ASSERT_EQ(1, X509_cmp_time(nextupd, &connection_time));
 
-  ASN1_GENERALIZEDTIME *revtime, *thisupd, *nextupd;
-  // Actual verification of the response
-  const int ocsp_resp_find_status_res = OCSP_resp_find_status(basic_response.get(), cert_id.get(), &status, &reason, &revtime, &thisupd, &nextupd);
-  ASSERT_EQ(1, ocsp_resp_find_status_res);
-  ASSERT_EQ(V_OCSP_CERTSTATUS_GOOD, status);
+  // Check valid connection timestamp right before OCSP response expires.
+  connection_time = valid_before_ocsp_expire_time;
+  ASSERT_EQ(-1, X509_cmp_time(thisupd, &connection_time));
+  ASSERT_EQ(1, X509_cmp_time(nextupd, &connection_time));
+
+  // Check expired connection timestamp
+  connection_time = invalid_after_ocsp_expire_time;
+  ASSERT_EQ(-1, X509_cmp_time(thisupd, &connection_time));
+  ASSERT_EQ(-1, X509_cmp_time(nextupd, &connection_time));
 }
 
 // Test against same good OCSP response, but checking behavior of not specifying
 // hash algorithm used for |OCSP_cert_to_id| this time (should default to sha1).
 TEST(OCSPTest, TestDefaultHash) {
   bssl::UniquePtr<OCSP_BASICRESP> basic_response;
-  ExtractBasicOCSP(ocsp_response_der, OCSP_RESPONSE_STATUS_SUCCESSFUL, &basic_response);
-  ASSERT_TRUE(basic_response);
+  bssl::UniquePtr<STACK_OF(X509)> server_cert_chain;
+  ExtractAndVerifyBasicOCSP(ocsp_response_der,
+                            OCSP_RESPONSE_STATUS_SUCCESSFUL,
+                            ca_cert,
+                            server_cert,
+                            1,
+                            &basic_response,
+                            &server_cert_chain);
 
-  bssl::UniquePtr<STACK_OF(X509)> server_cert_chain = CertsToStack(
-      {LoadX509fromPEM(server_cert).get(),LoadX509fromPEM(ca_cert).get()});;
-
-  X509 *subject = sk_X509_value(server_cert_chain.get(), 0);
-  X509 *issuer = sk_X509_value(server_cert_chain.get(), 1);
+  // Testing behavior of default hash algorithm, when |*dgst| is set to NULL.
+  // The hash algorithm should automatically be set to sha1. The revocation
+  // status check of the response should work if hash algorithm of |cert_id| has
+  // been set to sha1 successfully.
+  int status = 0;
+  ASN1_GENERALIZEDTIME *thisupd, *nextupd;
+  CheckOCSP_CERTSTATUS(&basic_response,
+                       &server_cert_chain,
+                       nullptr,
+                       1,
+                       &status,
+                       &thisupd,
+                       &nextupd);
+  ASSERT_EQ(V_OCSP_CERTSTATUS_GOOD, status);
+}
 
-  // Expect basic verify here, but we skip step for now since functionality has
-  // not been implemented yet.
+// Test valid OCSP response directly signed by the CA certificate
+TEST(OCSPTest, TestGoodOCSPbyCA) {
```

Comment:
Nit: some test code are similar (only the parameters, and expected value are different). The code can be unified by using Parameterized Tests `TEST_P`.

Reply:
I'll be working on a better abstraction in the next story `CryptoAlg-776` 👍 . After a closer look, I could do an abstraction with the tests that are highly similar, but the ones which have unique time checks and  data changes would most likely have to stay as standalone tests (eg `TestGoodOCSP` and `TestUntrustedDataOCSP`).